src/gtk/srchctrl.cpp

```diff
@@ -0,0 +1,429 @@
+/////////////////////////////////////////////////////////////////////////////
+// Name:        src/gtk/srchctrl.cpp
+// Purpose:     wxSearchCtrl implementation - native
+// Author:      Kettab Ali
+// Created:     2019-12-23
+// Licence:     wxWindows licence
+/////////////////////////////////////////////////////////////////////////////
+
+// For compilers that support precompilation, includes "wx.h".
+#include "wx/wxprec.h"
+
+#ifdef __BORLANDC__
+    #pragma hdrstop
+#endif
+
+#if wxUSE_SEARCHCTRL
+
+#include "wx/srchctrl.h"
+
+#ifndef WX_PRECOMP
+    #include "wx/menu.h"
+#endif //WX_PRECOMP
+
+#include "wx/utils.h"
+#include "wx/gtk/private.h"
+#include "wx/gtk/private/gtk3-compat.h"
+
+
+#if GTK_CHECK_VERSION(3,6,0)
+    // GtkSearchEntry is available only for GTK+ >= 3.6
+    #define wxHAS_GTK_SEARCH_ENTRY 1
+#else // GTK < 3.6
+    #define wxHAS_GTK_SEARCH_ENTRY 0
+#endif // GTK >= 3.6
+
+// ============================================================================
+// signal handlers implementation
+// ============================================================================
+
+extern "C" {
+
+static void
+wx_gtk_icon_press(GtkEntry* WXUNUSED(entry),
+                  gint position,
+                  GdkEventButton* WXUNUSED(event),
+                  wxSearchCtrl* ctrl)
+{
+    if ( position == GTK_ENTRY_ICON_PRIMARY )
+    {
+        // 1- Notice that contrary to the generic version, we don't generate the
+        //    wxEVT_SEARCH event here, which is the native behaviour of the
+        //    GtkSearchEntry (the search icon is inactive for a GtkSearchEntry).
+        //
+        // 2- If the wxSearchCtrl has a menu associated with it, we explicitly
+        //    make the search icon clickable as a way to display the menu.
+
+        ctrl->PopupSearchMenu();
+    }
+    else // position == GTK_ENTRY_ICON_SECONDARY
+    {
+        if ( !wx_is_at_least_gtk3(6) )
+        {
+            // No need to call this for a GtkSearchEntry.
+            ctrl->Clear();
+        }
+
+        wxCommandEvent event(wxEVT_SEARCH_CANCEL, ctrl->GetId());
+        event.SetEventObject(ctrl);
+        ctrl->HandleWindowEvent(event);
+    }
+}
+
+}
+
+// ============================================================================
+// wxSearchCtrl implementation
+// ============================================================================
+wxBEGIN_EVENT_TABLE(wxSearchCtrl, wxSearchCtrlBase)
+    EVT_CHAR(wxSearchCtrl::OnChar)
+    EVT_TEXT(wxID_ANY, wxSearchCtrl::OnText)
+    EVT_TEXT_ENTER(wxID_ANY, wxSearchCtrl::OnTextEnter)
+wxEND_EVENT_TABLE()
```

Comment:
Unless there is some compelling reason for this, I'd rather use `Bind()` in the new code. But well, it's not a big problem to leave the event table macros in neither, of course.